src/compiler/checker.ts

```diff
@@ -6814,37 +6814,58 @@ namespace ts {
             return undefined;
         }
 
-        function resolveTypeReferenceName(typeReferenceName: EntityNameExpression | EntityName) {
+        function resolveTypeReferenceName(node: TypeReferenceType, typeReferenceName: EntityNameExpression | EntityName) {
             if (!typeReferenceName) {
                 return unknownSymbol;
             }
 
-            return resolveEntityName(typeReferenceName, SymbolFlags.Type) || unknownSymbol;
+            const meaning = node.kind === SyntaxKind.JSDocTypeReference
+                ? SymbolFlags.Type | SymbolFlags.Value
+                : SymbolFlags.Type;
+
+            return resolveEntityName(typeReferenceName, meaning) || unknownSymbol;
         }
 
         function getTypeReferenceType(node: TypeReferenceType, symbol: Symbol) {
             const typeArguments = typeArgumentsFromTypeReferenceNode(node); // Do unconditionally so we mark type arguments as referenced.
+            let secondPass = true;
+            let fallbackType: Type = unknownType;
+            while (true) {

```

Comment:
Since this loop has a maximum of 2 iterations, couldn't you just define a function and call it twice?